File: testes/sagabackend/src/main/java/sistema/saga/sagabackend/model/Pessoa.java
package sistema.saga.sagabackend.model;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import sistema.saga.sagabackend.repository.PessoaDAO;

import java.time.LocalDate;
import java.util.List;

@Component
public class Pessoa {

    private String cpf;
    private String rg;
    private String nome;
    private LocalDate dataNascimento;
    private String sexo;
    private String locNascimento;
    private String estadoNascimento;
    private Endereco endereco;
    private String estadoCivil;

    @Autowired
    private PessoaDAO pessoaDAO;

    public Pessoa(String cpf, String rg, String nome, LocalDate dataNascimento, String sexo,
                  String locNascimento, String estadoNascimento, Endereco endereco, String estadoCivil) {
        this.cpf = cpf;
        this.rg = rg;
        this.nome = nome;
        this.dataNascimento = dataNascimento;
        this.sexo = sexo;
        this.locNascimento = locNascimento;
        this.estadoNascimento = estadoNascimento;
        this.endereco = endereco;
        this.estadoCivil = estadoCivil;
    }

    public Pessoa(String cpf) {
        this(cpf, "", "", null, "", "", "", null, "");
    }

    public Pessoa() {
        this("", "", "", null, "", "", "", null, "");
    }

    public String getCpf() {
        return cpf;
    }

    public void setCpf(String cpf) {
        this.cpf = cpf;
    }

    public String getRg() {
        return rg;
    }

    public void setRg(String rg) {
        this.rg = rg;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public LocalDate getDataNascimento() {
        return dataNascimento;
    }

    public void setDataNascimento(LocalDate dataNascimento) {
        this.dataNascimento = dataNascimento;
    }

    public String getSexo() {
        return sexo;
    }

    public void setSexo(String sexo) {
        this.sexo = sexo;
    }

    public String getLocNascimento() {
        return locNascimento;
    }

    public void setLocNascimento(String locNascimento) {
        this.locNascimento = locNascimento;
    }

    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    public Endereco getEndereco() {
        return endereco;
    }

    public void setEndereco(Endereco endereco) {
        this.endereco = endereco;
    }

    public String getEstadoCivil() {
        return estadoCivil;
    }

    public void setEstadoCivil(String estadoCivil) {
        this.estadoCivil = estadoCivil;
    }

    public boolean apagar() {
        return pessoaDAO.apagar(this);
    }

    public List<Pessoa> buscarTodos() {
        return pessoaDAO.get("");
    }

    public boolean alterar() {
        return pessoaDAO.alterar(this);
    }

    public boolean gravar() {
        return pessoaDAO.gravar(this);
    }
}

File: model/pessoa.js
import Endereco from "../model/endereco.js";
import PessoaDAO from "../repository/pessoaDAO.js";

export default class Pessoa {

    #cpf;
    #rg;
    #nome;
    #dataNascimento;
    #sexo;
    #locNascimento;
    #estadoNascimento;
    #estadoCivil;
    #endereco;

    get cpf() {
        return this.#cpf;
    }

    set cpf(novoCpf) {
        this.#cpf = novoCpf;
    }

    get rg() {
        return this.#rg;
    }

    set rg(novoRg) {
        this.#rg = novoRg;
    }

    get nome() {
        return this.#nome;
    }

    set nome(novoNome) {
        this.#nome = novoNome;
    }

    get dataNascimento() {
        return this.#dataNascimento;
    }

    set dataNascimento(novaDataNascimento) {
        this.#dataNascimento = novaDataNascimento;
    }

    get sexo() {
        return this.#sexo;
    }

    set sexo(novoSexo) {
        this.#sexo = novoSexo;
    }

    get locNascimento() {
        return this.#locNascimento;
    }

    set locNascimento(novoLocNascimento) {
        this.#locNascimento = novoLocNascimento;
    }

    get estadoNascimento() {
        return this.#estadoNascimento;
    }

    set estadoNascimento(novoEstadoNascimento) {
        this.#estadoNascimento = novoEstadoNascimento;
    }

    get estadoCivil() {
        return this.#estadoCivil;
    }

    set estadoCivil(novoEstadoCivil) {
        this.#estadoCivil = novoEstadoCivil;
    }

    get endereco() {
        return this.#endereco;
    }

    set endereco(novoEndereco) {
        if (novoEndereco instanceof Endereco) {
            this.#endereco = novoEndereco;
        } else {
            throw new Error("O endereço deve ser uma instância de Endereco.");
        }
    }

    constructor(cpf = "", rg = "", nome = "", dataNascimento = "", sexo = "", locNascimento = "", estadoNascimento = "", estadoCivil = "", endereco = new Endereco()) {
        this.#cpf = cpf;
        this.#rg = rg;
        this.#nome = nome;
        this.#dataNascimento = dataNascimento;
        this.#sexo = sexo;
        this.#locNascimento = locNascimento;
        this.#estadoNascimento = estadoNascimento;
        this.#estadoCivil = estadoCivil;
        this.#endereco = endereco;
    }

    toJSON() {
        return {
            "cpf": this.#cpf,
            "rg": this.#rg,
            "nome": this.#nome,
            "dataNascimento": this.#dataNascimento,
            "sexo": this.#sexo,
            "locNascimento": this.#locNascimento,
            "estadoNascimento": this.#estadoNascimento,
            "estadoCivil": this.#estadoCivil,
            "endereco": this.#endereco.toJSON()  
        };
    }

    async gravar() {
        const pessoaDAO = new PessoaDAO();
        return await pessoaDAO.gravar(this);
    }

    async alterar() {
        const pessoaDAO = new PessoaDAO();
        return await pessoaDAO.alterar(this);
    }

    async apagar() {
        const pessoaDAO = new PessoaDAO();
        return await pessoaDAO.apagar(this.#cpf);
    }

    async consultar(filtro) {
        const pessoaDAO = new PessoaDAO();
        return await pessoaDAO.get(filtro);
    }
}
